### DISCUSSION TITLE
Mutating Individual Building Blocks

### DISCUSSION THREAD

**Discussion by stevenkbennett** [lukasturcani/stk#291] [Ideas]
We want to think of a way to mutate individual building blocks as part of the EA for a project. Was wondering if this is worth implementing in stk? 

I had two thoughts about how this could be implemented. 
One idea could be generate these new building blocks when required in a `mutate` function. In this function, functional group substitutions could be performed on a randomly selected building block and a new ConstructedMolecule returned from these newly generated building blocks. 
The second idea I had was to somehow extend the original building block list with new ones as the EA progresses. This means existing mutation functions could be used to select from one of the newly appended building blocks.

Hope this makes sense!

**Comment by lukasturcani** [comment]
Yo, I've also got a graph based EA proposal here: it's work in progress and not directly related to this use-case but I figured its worth bringing up as an example of how the EA can be extended with more serious additions: https://github.com/lukasturcani/stk/issues/206

I think the first proposal makes sense, but it will only allow a limited set of mutations -> something like the graph based EA might allow for more extensive transformations.

I'm not sure how the second approach is different to using `RandomBuildingBlock`?

**Reply by stevenkbennett** [reply]
The graph-based EA sounds super interesting! I like the API you propose in #206 
Would be good to discuss with you some more ideas for this! 

I have a first go at the first idea of performing substitution in PR #300 with a SMARTS substructure replacer. It's quite a simple code, but I think can be useful convenience tool for users wanting do a quick replacement. 

I think you're correct actually that the second idea is identical to RandomBuildingBlock, maybe ignore that suggestion!
